db/crud.go

```diff
@@ -1862,7 +1864,12 @@ func (db *Database) updateAndReturnDoc(docid string, allowImport bool, expiry ui
 			Deleted:          doc.History[newRevID].Deleted,
 			_shallowCopyBody: storedDoc.Body(),
 		}
-		db.revisionCache.Put(documentRevision)
+
+		if revIDGenerationSkipped {
+			db.revisionCache.Upsert(documentRevision)
```

Comment:
We're upserting the rev cache after the document has been persisted to the bucket.  I think this means there's a new race here, in the (unlikely but possible) scenario where a client is notified of the new change and gets the previous rev from the rev cache? i.e.:
- document updated in bucket
- client gets notified of update, requests rev, gets previous version of rev 
- revisionCache.upsert runs
I think this race only creates a problem in the case when the user doesn't have security access to the previous version of the rev, but I don't see an obvious fix.  (even if we invalidate the old entry in the rev cache prior to writing the doc, another process could add it back)

Reply:
Addressed as per discussion.